### PR TITLE
coding challenge 5/17/2016

### PR DESCRIPTION
**Pull request by darwinlorenzopalonpon**

### CODE REVIEW

**Comment by darwinlorenzopalonpon** [comment]
- [x] 1. Use RSpec to specify Todo::Create
- [x] 2. Refactor Todo::Create

**Comment by kennethcalamay** [comment]
Hi Darwin, I'm KL from SS and I'll be taking a look at your code :-)

File: app/concepts/todo/create.rb [hunk 1]
```diff
@@ -1,50 +1,46 @@
...
-  # Logic for creating a TODO
-  #     if `todo_list` is specified
-  #         create the `todo`
-  #         associated with `todo_list`
-  #     else
-  #         create the `todo`
-  #         create a `user`
-  #         create a default `todo_list`
-  #         associate `todo_list` with `user`
-  #         associate `todo` with `todo_list`
```
Comment (kennethcalamay) [inline app/concepts/todo/create.rb:12]: Props for getting rid of the comments. 👍

File: app/concepts/todo/create.rb [hunk 2]
```diff
@@ -1,50 +1,46 @@
...
   def process(params)
-    # make sure title is not blank
-    return invalid! if params[:todo][:title].blank?
+    validate(params[:todo], @model) do |f|
+      todo_list = get_todo_list(params)
+      f.todo_list_id = todo_list.id
+      f.save
+    end
+  end
```
Comment (kennethcalamay) [inline app/concepts/todo/create.rb:16]: The process method looks clean.

File: app/concepts/todo/create.rb [hunk 3]
```diff
@@ -1,50 +1,46 @@
...
-
-      # assign the todo_list to the user
       todo_list.user = current_user
       todo_list.save!
-    else # i.e. if params[:todo_list_id] is not nil
+    else
       todo_list = TodoList.find(params[:todo_list_id])
     end
+    return todo_list
+  end
```
Comment (kennethcalamay) [inline app/concepts/todo/create.rb:30]: This could still be cleaned up.

File: app/concepts/todo/create.rb [hunk 4]
```diff
@@ -1,50 +1,46 @@
...
+    current_user =
+      if params[:current_user] == nil
+        if params[:current_user_id] == nil
+          User.create!(fullname: "Guest")
+        else
+          User.find(params[:current_user_id])
+        end
+      else
+        current_user
+      end
```
Comment (kennethcalamay) [inline app/concepts/todo/create.rb]: This could still be cleaned up.

File: spec/concepts/todo/create_spec.rb [hunk 5]
```diff
@@ -1,7 +1,31 @@
 require 'rails_helper'
 
 RSpec.describe Todo::Create do
-  it "can be spec'd" do
-    raise "Implement me"
+
+class Todo::Create
+  public :require
+end
```
Comment (kennethcalamay) [inline spec/concepts/todo/create_spec.rb]: This looks new to me. What does this do?

File: spec/concepts/todo/create_spec.rb [hunk 6]
```diff
@@ -1,7 +1,31 @@
...
+
+  context "when given a todo with no title" do
+    todo_params = ActionController::Parameters.new(todo: {title: nil})
```
Comment (kennethcalamay) [inline spec/concepts/todo/create_spec.rb:6]: Interesting usage of action controller parameters.

File: spec/concepts/todo/create_spec.rb [hunk 7]
```diff
@@ -1,7 +1,31 @@
...
+
+    it "should not create a todo item" do
+      res, op = Todo::Create.run(todo_params)
+      expect(op.model.id).to be_nil
+      expect(res).to be_falsey
+    end
```
Comment (kennethcalamay) [inline spec/concepts/todo/create_spec.rb:12]: Looks good.

File: spec/concepts/todo/create_spec.rb [hunk 8]
```diff
@@ -1,7 +1,31 @@
...
+
+  context "when given a todo with title" do
+    todo_params = ActionController::Parameters.new(todo: {title: "something"})
+    res, op = Todo::Create.run(todo_params)
+
+    it "should create a valid todo item" do
+      expect(op.model.id).not_to be_nil
+      expect(op.model.todo_list_id).not_to be_nil
+      expect(res).to be true
+    end
```
Comment (kennethcalamay) [inline spec/concepts/todo/create_spec.rb:32]: What about the persisted values?

**Comment by kennethcalamay** [comment]
Hi Darwin, your code looks good. It was easy to review. If you're still interested with us, can you demonstrate knowledge of separation of concerns / single responsibility principle both on method and class levels?

**Comment by darwinlorenzopalonpon** [comment]
Hi KL, 

I'll be checking out the comments and provide replies shortly, and then some code changes afterwards.

File: spec/concepts/todo/create_spec.rb (hunk 5 above)
Comment (darwinlorenzopalonpon) [inline spec/concepts/todo/create_spec.rb]: I ran into some issues with rails strong params (params.require() being a private method) when running the tests, so I used this patch to temporarily make the method public.

After a bit of reading on trailblazer, strong params can be done away with.

**Comment by darwinlorenzopalonpon** [comment]
- [x] Introduce separation of concerns / single responsibility
- [x] update spec to check for persisted values

**Comment by darwinlorenzopalonpon** [comment]
Hello KL,

I have updated the pull request.

**Pull request closed without merging**